libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp

```diff
@@ -0,0 +1,145 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AP_RangeFinder_config.h"
+
+#if AP_RANGEFINDER_JRE_SERIAL_ENABLED
+
+#include "AP_RangeFinder_JRE_Serial.h"
+#include <AP_Math/AP_Math.h>
+
+#define FRAME_HEADER_1    'R'    // 0x52
+#define FRAME_HEADER_2_A  'A'    // 0x41
+#define FRAME_HEADER_2_B  'B'    // 0x42
+#define FRAME_HEADER_2_C  'C'    // 0x43
+
+#define DIST_MAX_CM 5000
+#define OUT_OF_RANGE_ADD_CM   100
+
+void AP_RangeFinder_JRE_Serial::move_preamble_in_buffer(uint8_t search_start_pos)
+{
+    uint8_t i;
+    for (i=search_start_pos; i<data_buff_idx; i++) {
+        if (data_buff[i] == FRAME_HEADER_1) {
+            break;
+        }
+    }
+    if (i == 0) {
+        return;
+    }
+    memmove(data_buff, &data_buff[i], data_buff_idx-i);
+    data_buff_idx = data_buff_idx - i;
+}
+
+bool AP_RangeFinder_JRE_Serial::get_reading(float &reading_m)
+{
+    // uart instance check
+    if (uart == nullptr) {
+        return false;  // not update
+    }
+
+    uint16_t valid_count = 0;   // number of valid readings
+    uint16_t invalid_count = 0; // number of invalid readings
+    float sum = 0;
+    // max distance the sensor can reliably measure - read from parameters
+    const int16_t distance_cm_max = max_distance_cm();
+
+    // read a maximum of 8192 bytes per call to this function:
+    uint16_t bytes_available = MIN(uart->available(), 8192U);
+
+    while (bytes_available > 0) {
+        // fill buffer
+        const auto num_bytes_to_read = MIN(bytes_available, ARRAY_SIZE(data_buff) - data_buff_idx);
+        const auto num_bytes_read = uart->read(&data_buff[data_buff_idx], num_bytes_to_read);
+        if (num_bytes_read == 0) {
+            break;
+        }
+        if (bytes_available < num_bytes_read) {
+            // this is a bug in the uart call.
+            break;
+        }
+        bytes_available -= num_bytes_read;
+        data_buff_idx += num_bytes_read;
+
+        // move header frame header in buffer
+        move_preamble_in_buffer(0);
+
+        // ensure we have a packet type:
+        if (data_buff_idx < 2) {
+            continue;
+        }
+
+        // determine packet length for incoming packet:
+        uint8_t packet_length;
+        switch (data_buff[1]) {
+        case 'A':
```

Comment:
@peterbarker should we re-use the header definitions?  e.g. FRAME_HEADER_2_A.

Reply:
Just remove the header definitions IMO.

Unless you use the define in more than one place the only thing they add is an unnecessary level of indirection (i.e. to know what value you need to find in your diagnostic output you have to leave the code you're currently stepping through to find the `#define`.

Reply:
I don't feel too strongly about it but higher up in the move_preamble_in_buffer() function you've used the definition so I think I'll change these lines to also use the definition.  I think the definition names are slightly better than the magic characters simply appearing in the code although the same could probably be achieved by adding a comment.